Sources_v3/StreamChatUI/ChatChannelList/ChatChannelListViewController.swift

```diff
@@ -2,7 +2,116 @@
+    where ExtraData.Channel: NameAndImageProviding {
+    
+    // MARK: - Properties
+    
+    public let controller: _ChatChannelListController<ExtraData>
+    public let uiConfig: UIConfig<ExtraData>
+    public var didSelectChannel: (_ChatChannel<ExtraData>) -> Void = { _ in }
+    
+    // MARK: - Init
+
+    public init(
+        controller: _ChatChannelListController<ExtraData>,
+        uiConfig: UIConfig<ExtraData> = .default
+    ) {
+        self.controller = controller
+        self.uiConfig = uiConfig
+        
+        let layout = UICollectionViewFlowLayout()
+        layout.minimumInteritemSpacing = 0
+        layout.minimumLineSpacing = 0
+        
+        super.init(collectionViewLayout: layout)
+    }
+    
+    @available(*, unavailable)
+    public required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    // MARK: - Life Cycle
+    
+    override open func viewDidLoad() {
+        super.viewDidLoad()
+        
+        collectionView.backgroundColor = .white
+        collectionView.register(UICollectionViewCell.self, forCellWithReuseIdentifier: "Cell")
+        controller.setDelegate(self)
+        controller.synchronize()
+    }
+    
+    override open func viewWillLayoutSubviews() {
+        super.viewWillLayoutSubviews()
+        
+        (collectionView.collectionViewLayout as! UICollectionViewFlowLayout).itemSize = .init(
+            width: view.bounds.width,
+            height: 70
+        )
+    }
+    
+    // MARK: - UICollectionViewDataSource
+    
+    override public func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
+        controller.channels.count
+    }
+    
+    override public func collectionView(
+        _ collectionView: UICollectionView,
+        cellForItemAt indexPath: IndexPath
+    ) -> UICollectionViewCell {
+        let channel = controller.channels[indexPath.row]
+        
+        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "Cell", for: indexPath)
+        
+        if let channelView = cell.contentView.subviews.first as? ChatChannelView<ExtraData> {
+            channelView.channel = channel
+        } else {
+            let channelView = uiConfig.channelList.channelView.init(channel: channel, uiConfig: uiConfig)
+            channelView.translatesAutoresizingMaskIntoConstraints = false
+            cell.contentView.addSubview(channelView)
+            NSLayoutConstraint.activate([
+                channelView.leadingAnchor.constraint(equalTo: cell.contentView.leadingAnchor),
+                channelView.trailingAnchor.constraint(equalTo: cell.contentView.trailingAnchor),
+                channelView.topAnchor.constraint(equalTo: cell.contentView.topAnchor),
+                channelView.bottomAnchor.constraint(equalTo: cell.contentView.bottomAnchor)
+            ])
+        }
```

Comment:
The code for layout was previously inside custom `UICollectionViewCell` but that's not clear at this point whether the container cell should be generic over `ExtraData` as a content view is, or it should say just `content: UIView?`. 

Decided to go with default `UICollectionViewCell` and discuss it later on.

Reply:
added in 0ef92e65959e905564a5a1660f6fd50fc7ca2e15, used in b41f9781372edb9336a4c1739e19ba8b267c615a